### PR TITLE
CTCP-5480: Route details frontend - change country of transport

### PR DESCRIPTION
**Pull request by joecefnhedog**

### CODE REVIEW

File: app/controllers/routing/index/CountryOfRoutingController.scala [hunk 1]
```diff
@@ -72,24 +79,52 @@ class CountryOfRoutingController @Inject() (
...
+    page: Index => QuestionPage[CustomsOffice],
+    countryCode: String
+  ): Try[UserAnswers] =
+    (0 until userAnswers.get(array).length).foldRight(Try(userAnswers)) {
+      case (index, acc) =>
+        userAnswers.get(page(Index(index))) match {
+          case Some(value) if value.countryId == countryCode => acc.flatMap(_.remove(obj(Index(index))))
+          case _                                             => acc
+        }
+    }
```
Comment (jackheslop96) [inline app/controllers/routing/index/CountryOfRoutingController.scala:129]: Can this not just live inside CountryOfRoutingPage cleanup?

Comment (joecefnhedog) [reply app/controllers/routing/index/CountryOfRoutingController.scala:129]: We initially tried to put it in there, but we needed the previous country (to remove transitOffices related to that country) but that` value: Option[Country]` was the new selected country. Maybe there was a way to pass that in?

File: app/controllers/package.scala [hunk 2]
```diff
@@ -43,13 +46,14 @@ package object controllers {
   implicit class SettableOps[A](page: QuestionPage[A]) {
 
     def writeToUserAnswers(value: A)(implicit format: Format[A]): UserAnswersWriter[Write[A]] =
-      ReaderT[EitherType, UserAnswers, Write[A]](
+      ReaderT[EitherType, UserAnswers, Write[A]] {
         userAnswers =>
+          println("writeTo", userAnswers)
```
Comment (joecefnhedog) [inline app/controllers/package.scala:51]:
```suggestion

```

File: app/controllers/package.scala [hunk 3]
```diff
@@ -63,6 +66,30 @@ package object controllers {
 
   implicit class SettableOpsRunner[A](userAnswersWriter: UserAnswersWriter[Write[A]]) {
 
+    def removeOfficesOfTransit(previousSelectedCountry: Option[Country], selectedCountry: Country): UserAnswersWriter[Write[A]] =
+      userAnswersWriter.flatMapF {
+        case (page, userAnswers) =>
+          println("remPAge", page)
```
Comment (joecefnhedog) [inline app/controllers/package.scala:72]:
```suggestion
```

File: app/navigation/CountryOfRoutingNavigator.scala [hunk 4]
```diff
@@ -27,12 +27,8 @@ import javax.inject.{Inject, Singleton}
 class CountryOfRoutingNavigatorProviderImpl @Inject() (implicit config: FrontendAppConfig, phaseConfig: PhaseConfig) extends CountryOfRoutingNavigatorProvider {
 
   def apply(mode: Mode, index: Index): UserAnswersNavigator =
-    mode match {
-      case NormalMode =>
-        new CountryOfRoutingNavigator(mode, index)
-      case CheckMode =>
-        new RouteDetailsNavigator(mode)
-    }
+    new CountryOfRoutingNavigator(mode, index)
```
Comment (jackheslop96) [inline app/navigation/CountryOfRoutingNavigator.scala:30]: Hmm...not sure about this?

File: app/models/journeyDomain/routing/CountryOfRoutingDomain.scala [hunk 5]
```diff
@@ -34,7 +35,11 @@ case class CountryOfRoutingDomain(
       case AccessingJourney =>
         Some(controllers.routing.index.routes.CountryOfRoutingController.onPageLoad(userAnswers.lrn, mode, index))
       case CompletingJourney =>
-        CountriesOfRoutingSection.route(userAnswers, mode)
+        mode match {
+          case NormalMode => CountriesOfRoutingSection.route(userAnswers, mode)
+          case CheckMode  => RouteDetailsSection.route(userAnswers, mode)
+        }
```
Comment (jackheslop96) [inline app/models/journeyDomain/routing/CountryOfRoutingDomain.scala:41]: I think this can be reverted along with the `CountryOfRoutingNavigator` change, but would be good to understand why this change was made?

File: app/pages/routing/index/CountryOfRoutingPage.scala [hunk 6]
```diff
@@ -35,12 +34,9 @@ case class CountryOfRoutingPage(index: Index) extends QuestionPage[Country] {
...
   override def route(userAnswers: UserAnswers, mode: Mode): Option[Call] =
-    Some(routes.CountryOfRoutingController.onPageLoad(userAnswers.lrn, mode, index))
-
-  override def cleanup(value: Option[Country], userAnswers: UserAnswers): Try[UserAnswers] =
-    value match {
-      case Some(_) => userAnswers.remove(TransitSection).flatMap(_.remove(ExitSection))
-      case None    => super.cleanup(value, userAnswers)
+    mode match {
+      case NormalMode => Some(routes.CountryOfRoutingController.onPageLoad(userAnswers.lrn, mode, index))
+      case CheckMode  => Some(controllers.routes.RouteDetailsAnswersController.onPageLoad(userAnswers.lrn))
```
Comment (jackheslop96) [inline app/pages/routing/index/CountryOfRoutingPage.scala:39]: I think this can be reverted along with the `CountryOfRoutingNavigator` change, but would be good to understand why this change was made?

File: app/controllers/routing/index/CountryOfRoutingController.scala [hunk 7]
```diff
@@ -72,22 +72,25 @@ class CountryOfRoutingController @Inject() (
             .bindFromRequest()
             .fold(
               formWithErrors => Future.successful(BadRequest(view(formWithErrors, lrn, countryList.values, mode, index))),
-              value =>
+              selectedCountry =>
```
Comment (jackheslop96) [inline app/controllers/routing/index/CountryOfRoutingController.scala:75]: Sorry. OCD. For the sake of consistency can we just call this value?

File: app/controllers/package.scala [hunk 8]
```diff
@@ -63,6 +68,31 @@ package object controllers {
...
+                (page,
+                 userAnswers
+                   .findAndRemoveOffices(OfficesOfTransitSection,
+                                         OfficeOfTransitSection,
+                                         OfficeOfTransitPage,
+                                         previousCountry.code.code,
+                                         selectedCountry.code.code
+                   )
+                   .findAndRemoveOffices(OfficesOfExitSection, OfficeOfExitSection, OfficeOfExitPage, previousCountry.code.code, selectedCountry.code.code)
+                )
```
Comment (jackheslop96) [inline app/controllers/package.scala:86]: Again, me being OCD haha, can we format these two `findAndRemoveOffices` calls the same way
